pandas/parser.pyx

```diff
@@ -217,6 +221,24 @@ cdef extern from "parser/tokenizer.h":
     int to_boolean(const char *item, uint8_t *val) nogil
 
 
+# XXX
+# this is a hack - in order to make the inference
+# functions generic (converting either data directly
+# from the parser or from a passed in hash table)
+# we add an "optional" parameter via fused type, that can either
+# be the hash table to parse, or an integer, which is used
+# as a sentinel to specialize the function for reading
+# from the parser.
+
+# This is to avoid duplicating a bunch of code or
```

Comment:
The type inference functions (e.g. `_try_double`) read data straight from the parser -  the `COLITER_NEXT(it,word)` stuff.  I wanted to also be able to pass in a hash table to that function, i.e. my categories, and this was my workaround.